src/aten/sycl/SortingKernels.h

```diff
@@ -0,0 +1,508 @@
+          end_bit,
+          count);
+  sycl_kernel_submit(
+      num_segments * num_tiles * GROUP_SIZE,
+      GROUP_SIZE,
+      at::xpu::getCurrentSYCLQueue(),
+      caller);
+}
+
+// ======================= large sort =======================
+
+template <
+    typename key_t,
+    typename value_t,
+    bool IS_DESCENDING,
+    int KEYS_PER_ITEM,
+    int GROUP_SIZE,
+    int SUBGROUP_SIZE>
+void segmented_radix_sort_pairs_kernel(
+    const key_t* keys_in,
+    key_t* keys_out,
+    const value_t* values_in,
+    value_t* values_out,
+    int num_segments,
+    int num_elements) {
+  constexpr int TILE_PROCESSING_LENGTH = GROUP_SIZE * KEYS_PER_ITEM;
+  int num_tiles =
+      (num_elements + TILE_PROCESSING_LENGTH - 1) / TILE_PROCESSING_LENGTH;
+  constexpr int RADIX_BITS = 4;
+  constexpr int RADIX_BUCKETS = 16;
+  int begin_bit = 0;
+  int end_bit = KeyTraits<key_t>::endbit();
+  int* counts;
+  key_t* keys_temp;
+  value_t* values_temp;
+
+  at::DataPtr counts_data = c10::xpu::XPUCachingAllocator::get()->allocate(
+      num_segments * RADIX_BUCKETS * num_tiles * sizeof(int));
+  at::DataPtr keys_temp_data = c10::xpu::XPUCachingAllocator::get()->allocate(
+      num_segments * num_elements * sizeof(key_t));
+  at::DataPtr values_temp_data = c10::xpu::XPUCachingAllocator::get()->allocate(
+      num_segments * num_elements * sizeof(value_t));
+
+  counts = (int*)counts_data.get();
+  keys_temp = (key_t*)keys_temp_data.get();
+  values_temp = (value_t*)values_temp_data.get();
+
+  key_t* keys_in_ = const_cast<key_t*>(keys_in);
+  key_t* keys_out_ = keys_temp;
+  value_t* values_in_ = const_cast<value_t*>(values_in);
+  value_t* values_out_ = values_temp;
+
+  while (true) {
+    segmented_radix_sort_pairs_upsweep_kernel<
+        key_t,
+        value_t,
+        IS_DESCENDING,
+        KEYS_PER_ITEM,
+        GROUP_SIZE,
+        SUBGROUP_SIZE>(
+        keys_in_, counts, num_segments, num_elements, begin_bit, end_bit);
+
+    segmented_radix_sort_pairs_scan_kernel<
+        KEYS_PER_ITEM,
+        GROUP_SIZE,
+        SUBGROUP_SIZE>(counts, num_tiles, num_segments);
+
+    segmented_radix_sort_pairs_downsweep_kernel<
+        key_t,
+        value_t,
+        IS_DESCENDING,
+        KEYS_PER_ITEM,
+        GROUP_SIZE,
+        SUBGROUP_SIZE>(
+        keys_in_,
+        keys_out_,
+        values_in_,
+        values_out_,
+        num_segments,
+        num_elements,
+        begin_bit,
+        end_bit,
+        counts);
+
+    if (begin_bit == 0) {
+      keys_in_ = keys_temp;
+      keys_out_ = keys_out;
+      values_in_ = values_temp;
+      values_out_ = values_out;
+    } else {
+      std::swap(keys_in_, keys_out_);
+      std::swap(values_in_, values_out_);
+    }
+    begin_bit += RADIX_BITS;
+    if (begin_bit >= end_bit)
+      break;
+  }
+}
+
+// ======================= interface =======================
+
+template <
+    typename key_t,
+    typename value_t,
+    bool IS_DESCENDING,
+    int SUBGROUP_SIZE = 32>
```

Comment:
Please add comments for hard-coded sg size. It is not platform general code. We say we agree some specific code for kernel performance on specific hardware platform. But we need some clear comments for it.